Visualization utilities

svank opened the pull request:
Here's a more substantive PR, involving design decisions---I expect comments and opinions! :)

I've rigged up `find_stars_and_average` to store the number of stars contributing to each patch, and to store those as a `.counts` attribute on the `PatchCollection`.

I added a `visualization` module with two functions---one to plot that `counts` data:
```
plt.figure(figsize=(7,7))
regularizepsf.visualize.visualize_patch_counts(patch_collection, ax=plt.gca())
```
![image](https://user-images.githubusercontent.com/23462789/229225003-eb217eef-5176-4bd1-91be-3f400163a644.png)

For this plot, at first I was making an array containing the counts and `imshow`ing that, to make a plot like this (color-by-numbers!):
```
--------------------
| 7 | 12 | 11 | 5  |
| 6 | 8  | 19 | 10 |
--------------------
```
But the patches overlap each other and stars contribute to multiple patches, so I thought that sort of plot with space-filling array cells makes it really easy to mis-interpret the plot as marking the number of stars contained within each cell. So instead I did this scatter-plot approach.

The second function plots the patches themselves. (You must also have code to do this---maybe you should substitute yours in here)
```
regularizepsf.visualize.visualize_patches(cpc, psf_size=psf_size, figsize=(12,12), uniform_scaling=True)
```
![image](https://user-images.githubusercontent.com/23462789/229224054-c0b54056-ee55-4d6c-893e-86709c72cb20.png)

Of note, none of this has tests currently

jmbhughes commented:
I haven't looked at your code yet, but I will by the end of Monday. 

As you suggested, I might merge some of my patch plotting code into yours. I might also add a function to plot both a before and after since that's a comparison that's nice to have. 
![punch](https://user-images.githubusercontent.com/13068381/229236496-2fedbe72-d30f-43de-8813-6e18df62cb87.png)

As for tests, I'm concerned how this pulls down test coverage. Writing good unit tests for Matplotlib code can be tricky. A quick Google search gives many suggestions on how to unit test figure generation code. Even just testing that it runs would be better than nothing.

jmbhughes commented:
Also, I think we need to add `matplotlib` to the `setup.py` `install_requires` list. Probably should validate that contains everything in `requirements.txt` too.

svank commented:
Your patch plot looks better than mine.

> Annoyingly, this plot has a transparent background that doesn't render well in GitHub's dark mode.

Jupyter? If you drop `c.InlineBackend.rc = {'figure.facecolor': 'white'}` in `~/.ipython/profile_default/ipython_kernel_config.py`, matplotlib will fill plots in with a solid white background (and I wish this were the default!)

I have some matplotlib-based tests in https://github.com/svank/wispr_analysis, but they feel really flakey---especially anything that includes text, since apparently font rendering is incredibly platform-specific

jmbhughes commented on regularizepsf/visualize.py:
```diff
@@ -0,0 +1,125 @@
+    return ax
+
+
+def visualize_patches(patch_collection: PatchCollectionABC,
```
Here's the code I used to visualize a before and after set of patches:
```python
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
import itertools
import matplotlib.pylab as pylab
import matplotlib.colors as colors
params = {'axes.titlesize': 15}
pylab.rcParams.update(params)

vmin = 0.02
vmax = 1.0

fig = plt.figure(constrained_layout=True, figsize=(13, 6))

gs = GridSpec(4, 9, figure=fig, 
              width_ratios=([1]*4) + [0.8] + ([1]*4),
              height_ratios=[1]*4)

uncorrected_axs = dict()
for i, j in itertools.product(range(4), range(4)):
    uncorrected_axs[i, j] = fig.add_subplot(gs[i, j])
    
for i in range(16):
    row, col = i % 4, i // 4
    x = coords[np.array([1, 3, 5, 7])][row]
    y = coords[np.array([1, 3, 5, 7])][col]
    im = uncorrected_axs[3-row, col].imshow(array_corrector[x, y][117:-117, 117:-117], 
                                  origin='lower', cmap=custom, 
                                            norm=colors.PowerNorm(gamma=1/2.2, vmin=vmin, vmax=vmax))
    uncorrected_axs[3-row, col].set_axis_off()
    uncorrected_axs[3-row, col].set_aspect(1)
    
cax = fig.add_axes([0.479, 0.065, 0.012, 0.85])
#fig.colorbar(im, cax=cax, orientation='vertical', label='log$_{10}$(brightness)')
fig.colorbar(im, cax=cax, orientation='vertical', label='brightness')

corrected_axs = dict()
for i, j in itertools.product(range(4), range(4)):
    corrected_axs[i, j] = fig.add_subplot(gs[i, j+5])
for i in range(16):
    row, col = i // 4, i % 4
    x = coords[np.array([1, 3, 5, 7])][row]
    y = coords[np.array([1, 3, 5, 7])][col]
    im = corrected_axs[3-row, col].imshow(output[CoordinateIdentifier(None, x, y)][117:-117, 117:-117], 
                                    origin='lower', cmap=custom, 
                                          norm=colors.PowerNorm(gamma=1/2.2, vmin=vmin, vmax=vmax))
    corrected_axs[3-row, col].set_axis_off()
    corrected_axs[3-row, col].set_aspect(1)

    
plt.figtext(0.23, 0.95, 'Uncorrected', ha='center', va='center', fontsize=15)
plt.figtext(0.23+0.54, 0.95, 'Corrected', ha='center', va='center', fontsize=15)
uncorrected_axs[0, 2].set_title("\n")
plt.show()
fig.savefig("/Users/jhughes/Desktop/punch.png", dpi=300)
```

It would obviously need generalizing if we were to add it.

svank replied:
> It would obviously need generalizing if we were to add it.

`fig.savefig("/Users/jhughes/Desktop/punch.png", dpi=300)`

I think we can just require the user to log in as `jhughes`, no need to generalize

svank replied:
But I do prefer your patch plots, so I think this code should be the base of what we end up on

jmbhughes commented on regularizepsf/fitter.py:
```diff
@@ -22,8 +22,9 @@
 
 class PatchCollectionABC(metaclass=abc.ABCMeta):
-    def __init__(self, patches: dict[Any, np.ndarray]) -> None:
+    def __init__(self, patches: dict[Any, np.ndarray], counts: dict[Any, int] = None) -> None:
```
I think the type hint of `counts` should be `Optional` since we're setting it to `None`.

jmbhughes commented:
We should make a page in the Jupyter-book docs website for this module.

jmbhughes commented:
As for testing, it's not absolutely necessary. I think we can just say in any software reviews that testing plotting is finicky. 88% isn't bad. (Maybe we could boost it a little by checking if there's other code not covered.)  

It [looks like you can specify to not show text](https://matplotlib.org/stable/devel/testing.html#writing-an-image-comparison-test) when running a Matplotlib test. Maybe that would help?

svank commented on regularizepsf/visualize.py:
```diff
@@ -0,0 +1,125 @@
+    uniform_scaling : boolean
+        If True, use the scame colormap scaling for all patches. If False,
+        normalize each one separately.
+    imshow_args : dict
```
Ah, I was looking through the default matplotlib colormaps and couldn't figure out which one you were using---this is why! And the gamma correction is probably a good default

svank commented on regularizepsf/visualize.py:
```diff
@@ -0,0 +1,125 @@
+            patch = patch[trim:-trim, trim:-trim]
+        # Since the patches overlap each other, we need to plot each one so it
+        # only reaches out to the point of overlap.
+        extent = (corner.y + patch_size/4,
```
I think it should be possible both to plot a fraction of the patches, and to plot all the patches. The former may be good for cleaner illustrations, while the latter is good for checking that all the patches look reasonable.

One thing I was unsure about in my code was plotting all the patches (from all four grids) together inside one large set of axes labeled with pixel numbers. The axis labels help connect a patch to an image location, but they misrepresent the range of each patch, since I was cramming them all into one big plot despite their sample overlapping regions. I like that your approach removes all the axis labels and adds small borders between each patch, to make their locations more representative than pixel-perfect.

svank commented:
I've updated to roll in your patch-plotting code.

I have a before-and-after mode (in this demo the "corrected" side isn't actually corrected, because I don't have that ready yet):
`regularizepsf.visualize.visualize_PSFs(array_corrector, cpc, False, region_size=psf_size)`
![image](https://user-images.githubusercontent.com/23462789/230445546-415fe14b-ce6a-4485-a778-9426b7a39f75.png)

and a plain "what's my PSF" mode:
`regularizepsf.visualize.visualize_PSFs(array_corrector, None, False, region_size=psf_size)`
![image](https://user-images.githubusercontent.com/23462789/230445658-6d139076-2f2c-475b-8390-5ba9aced2372.png)

Both modes can also plot all the patches:
`regularizepsf.visualize.visualize_PSFs(array_corrector, None, True, region_size=psf_size)`
![image](https://user-images.githubusercontent.com/23462789/230445751-dc4e302d-6b03-4f3e-95c1-8dc9fea64b26.png)

svank commented:
I've changed `visualize_patch_counts` to `imshow` an array of count values, rather than make a scatter plot. It makes for a cleaner plot, I think, and doesn't result in overlapping scatterplot dots when there's a large number of patches. My previous reason for not doing it this way was that the axis labels didn't really make sense as pixel numbers, but I've changed it to label the axes as just "patch number" instead.
![image](https://user-images.githubusercontent.com/23462789/230653508-beb7ad17-0ae0-4c18-8e30-5113357c9cdf.png)

I've also added first-draft tests for these two functions. I used `pytest-mpl`, which has the advantage over the `matplotlib.testing` tools that it doesn't save out the generated plots when the tests pass. It doesn't run the image comparison tests at all unless you run the tests as `pytest --mpl`, and there's another flag for re-generating the reference images, so I added a stub "Development" page to the docs describing this. I think having the image tests not run by default is so that you can restrict them to a consistent test environment with consistent font rendering. That can be worked around by removing all text from the plots---I was disappointed that the `remove_text` option only removes *some* text, so I went heavy on removing all text from the image tests.

svank commented:
Just continuing to add more for you to look at when you can, I added a tool to compute and plot the transfer kernel corresponding to each patch. Looks like this:
`regularizepsf.visualize.visualize_transfer_kernels(array_corrector, alpha, epsilon, region_size=psf_size)`
![image](https://user-images.githubusercontent.com/23462789/231301048-4bc15ad5-34ac-4a55-941d-615fa715509f.png)

To make this work easily, I went into the Cython code and pulled the regularized inverse into its own function.

Something that occurred to me is that for the image-comparison tests I added, the reference images are generated with the changes of #19 and #28 in effect, so the tests won't pass as-is unless this PR is applied on top of those.

svank commented:
Another update I'm proud of: labeling the axes of these plots with the pixel bounds of each patch. I found this really helpful for matching up these plots to the actual data. It makes the axes a bit more verbose, though, so I have this feature default to off---for large images or small patch sizes, there could be ugly overlapping text.

I tried two styles, and should probably choose just one for everything. The second is a bit cleaner but takes up more space.

![image](https://user-images.githubusercontent.com/23462789/231594943-8753ee97-294f-429a-9779-e823f77bc5d3.png)

![image](https://user-images.githubusercontent.com/23462789/231594955-920b0415-7d2d-40c5-be4d-d0617f054d66.png)

jmbhughes commented:
This looks good. I think I like the second one better too.

jmbhughes commented:
@svank I went ahead and merged the other PRs.

svank commented:
I've added docs---I think this PR is in a mergeable state now

jmbhughes commented:
Awesome! Thanks for adding all of this!
